src/control/cmd/agent/infocache.go

```diff
@@ -0,0 +1,137 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+package main
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/pkg/errors"
+
+	mgmtpb "github.com/daos-stack/daos/src/control/common/proto/mgmt"
+	"github.com/daos-stack/daos/src/control/drpc"
+	"github.com/daos-stack/daos/src/control/lib/atm"
+	"github.com/daos-stack/daos/src/control/lib/netdetect"
+	"github.com/daos-stack/daos/src/control/logging"
+)
+
+const defaultNumaNode = 0
+
+type attachInfoCache struct {
+	log logging.Logger
+	// is caching enabled?
+	enabled atm.Bool
+	// is the cache initialized?
+	initialized atm.Bool
+	// maps NUMA affinity and device index to a response
+	resmgmtpb map[int]map[int][]byte
+	// maps NUMA affinity to a device index
+	devIdx map[int]int
+	mutex  sync.Mutex
+}
+
+// loadBalance is a simple round-robin load balancing scheme
+// to assign network interface adapters to clients
+// on the same NUMA node that have multiple adapters
+// to choose from.  Returns the index of the device to use.
+func (aic *attachInfoCache) loadBalance(numaNode int) int {
+	aic.mutex.Lock()
+	numDevs := len(aic.resmgmtpb[numaNode])
+	// keep the deviceIndex in range of the current resmgmtpb map
+	deviceIndex := aic.devIdx[numaNode] % numDevs
+	aic.devIdx[numaNode] = (deviceIndex + 1) % numDevs
+	aic.mutex.Unlock()
+	return deviceIndex
+}
+
+func (aic *attachInfoCache) getResponse(numaNode int) ([]byte, error) {
+	deviceIndex := aic.loadBalance(numaNode)
+	resmgmtpb, ok := aic.resmgmtpb[numaNode][deviceIndex]
+	if !ok {
+		return nil, errors.Errorf("GetAttachInfo entry for numaNode %d device index %d did not exist", numaNode, deviceIndex)
+	}
+	aic.log.Debugf("Retrieved response for NUMA %d with device index %d\n", numaNode, deviceIndex)
+	return resmgmtpb, nil
+}
+
+// initResponseCache generates a unique dRPC response corresponding to each device specified
+// in the scanResults.  The responses are differentiated based on the network device NUMA affinity.
+func (aic *attachInfoCache) initResponseCache(resp *mgmtpb.GetAttachInfoResp, scanResults []netdetect.FabricScan) error {
+	aic.mutex.Lock()
+	defer aic.mutex.Unlock()
```

Comment:
There's quite a lot of heavyweight (i.e. long-running) code under this lock, but I guess the cache is only being initialized on the first client dRPC request, right? Would it make sense to call this on agent startup instead? 

I'm a little concerned about two issues here: 
- First, a "thundering herd" problem where a massively distributed application starts on nodes with fresh agents and then every single compute node hammers the AP at the same time. If it's not a big deal to move this to agent startup, I think that would best allow us to evaluate further optimizations at a later time (e.g. putting a random delay from startup before the cache initialization, etc).
- Second, if we wait until the first request to initialize the cache, then we may be needlessly delaying an application's startup if it happens to be the unlucky first application to hit the agent. I may be overthinking it, but I can imagine a future support request to understand why some jobs start more slowly than others.

Reply:
Based on our sideband conversation, looks like we agree to initialize the cache on the first client dRPC request as it is now, rather than pre-fetch the data and fill the cache at agent startup.